scripts/doc_checker.py

```diff
@@ -0,0 +1,517 @@
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""
+Validate docstrings using pydocstyle and numpydoc.
+
+Example usage:
+python scripts/doc_checker.py asv_bench/benchmarks/utils.py modin/pandas
+"""
+
+import argparse
+import pathlib
+import subprocess
+import os
+import re
+import ast
+from typing import List
+import sys
+import inspect
+from numpydoc.validate import Docstring
+
+MODIN_PATH = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+sys.path.insert(0, MODIN_PATH)
+
+# error codes that pandas test in CI
+# https://numpydoc.readthedocs.io/en/latest/validation.html#built-in-validation-checks
+NUMPYDOC_BASE_ERROR_CODES = {
+    *("GL01", "GL02", "GL03", "GL05", "GL06", "GL07", "GL08", "GL09", "GL10"),
+    *("SS02", "SS03", "SS04", "SS05", "PR01", "PR02", "PR03", "PR04", "PR05"),
+    *("PR08", "PR09", "PR10", "RT01", "RT04", "RT05", "SA02", "SA03"),
+}
+
+MODIN_ERROR_CODES = {
+    "MD01": "'{parameter}' description should be '[type], default: [value]', found: '{found}'",
+    "MD02": "Spelling error in line: {line}, found: '{word}', reference: '{reference}'",
+}
+
+
+def get_optional_args(doc: Docstring) -> dict:
+    """
+    Get optional parameters for the object for which the docstring is checked.
+
+    Parameters
+    ----------
+    doc : numpydoc.validate.Docstring
+        Docstring handler.
+
+    Returns
+    -------
+    dict
+        Dict with default argument names and its values.
+    """
+    obj = doc.obj
+    if not callable(obj) or inspect.isclass(obj):
+        return {}
+    signature = inspect.signature(obj)
+    return {
+        k: v.default
+        for k, v in signature.parameters.items()
+        if v.default is not inspect.Parameter.empty
+    }
+
+
+def check_optional_args(doc: Docstring) -> list:
+    """
+    Check type description of optional arguments.
+
+    Parameters
+    ----------
+    doc : numpydoc.validate.Docstring
+
+    Returns
+    -------
+    list
+        List of tuples with Modin error code and its description.
+    """
+    if not doc.doc_parameters:
+        return []
+    optional_args = get_optional_args(doc)
+    if not optional_args:
+        return []
+
+    errors = []
+    for parameter in optional_args:
+        type_line = doc.doc_parameters[parameter][0]
+        if "default: " not in type_line or "optional" in type_line:
```

Comment:
```suggestion
        has_default = "default: " in type_line
        has_optional = "optional" in type_line
        if not (has_default ^ has_optional):
```
I think it's a missing `not` in here.
Plus it doesn't check that we specify both, while we should pick one of the two (which is a definition of "exclusive OR" operation).